### DISCUSSION TITLE
set status and/or message for failed workflows

### DISCUSSION THREAD

**Discussion by static-moonlight** [argoproj/argo-workflows#13084] [Q&A]
I have workflows which can fail in certain situations, however, there is no point in keeping them around. Analysis of missing data happens later on anyway. Plus, I put entries in a status database and export Prometheus metrics to make it visible.
Is there a way to use an exit handler to ...
- ... override the status/phase for a workflow to mark it as successful (even, if some tasks failed)?
- ... override the message (in this case: "missing data"), so that it is easy in the ui to know why they failed?
- ... override the `secondsAfterFailure` for this particular workflow and set it to `0` so that the workflow is being removed sooner?

**Comment by Joibel** [comment]
* you can use [continueOn](https://argo-workflows.readthedocs.io/en/latest/fields/#continueon) on tasks or steps to ignore their result.
* how are you causing the message `missing data`... that might tell you how to override the message...
* you can set a ttlStrategy in the [workflow.spec](https://argo-workflows.readthedocs.io/en/latest/fields/#workflowspec)

**Comment by static-moonlight** [comment]
Let me try again ...

Normally, I keep failed workflows for 1 week, so that somebody can take a look at them. That means, I already know about the ttl strategy because I'm using it to set `secondsAfterFailure` to 1 week. However, some of those failed workflows are just unable to recover or repeat. Simply put, there is nothing anyone can do about them, because some fundamental data is missing for good. If that happens, I don't want or need to keep the failed workflow for 1 week in my Argo system. So I want to put in an error handling into my workflow to automatically let Argo remove those particular workflows right away. An operator is not supposed to deal with such scenarios, because they can't fix it anyway. The only thing they can do is to delete that workflow manually (i just  want to automate that part).

I don't see how using `continueOn` would help with that. I assume it will just cause more workflow steps to fail, because one required step has already failed, and everything after that will fail too. In the end, my workflow would still be failed and kept for 1 week.

> how are you causing the message missing data... that might tell you how to override the message...

I don't understand that answer. I could identify the state "missing data" by checking the exit code of one particular step. So, in case I can't get rid of the workflow right away, alternatively I'd like to show a custom message right here:

![workflow](https://github.com/argoproj/argo-workflows/assets/107991124/603b524b-d2e8-40d3-acb4-f705f0a023cd)

Is that possible? If yes, how?

Simply put, I need one of those:
- if exit code of workflow step "get data" is 123, then set workflow status to successful
- if exit code of workflow step "get data" is 123, then set `secondsAfterFailure` for this workflow to `0`
- if exit code of workflow step "get data" is 123, then set workflow message to "missing data"
- if exit code of workflow step "get data" is 123, then add label to workflow `status=missing-data` (least preferable option, but still interesting to know)

I am also open for alternative suggestions how to deal with such a scenario.

**Reply by jswxstw** [reply]
> if exit code of workflow step "get data" is 123, then set workflow message to "missing data"

You may try to write "missing data" to "/dev/termination-log" in workflow step "get data" when it runs failed with expected error.
The error message "missing data" will be exposed to the workflow.

**Reply by static-moonlight** [reply]
> You may try to write "missing data" to "/dev/termination-log" in workflow step "get data" when it runs failed with expected error. The error message "missing data" will be exposed to the workflow.

That doesn't seem to work. I see this in the MESSAGE field for the task:

![image](https://github.com/argoproj/argo-workflows/assets/107991124/0d7e7112-e9ad-44a4-9ce5-de782fb439b9)

... but the MESSAGE for the workflow is still just `-`

![image](https://github.com/argoproj/argo-workflows/assets/107991124/d6842dae-3e6c-449d-9ba2-178494cc41b0)

**Reply by jswxstw** [reply]
> ... but the MESSAGE for the workflow is still just `-`

https://github.com/argoproj/argo-workflows/blob/a681b3bd73d4d8d01be697af06b18594b94180ca/workflow/controller/operator.go#L482
Workflow message is usually the message from the first failed node, which is typically a StepGroup type node rather than a Pod type node, if Steps are used. Unfortunately, it can not meet your requirements.
<img width="1295" alt="image" src="https://github.com/argoproj/argo-workflows/assets/18543532/ab0459b7-7fc5-4a3c-bad9-1ae12f7f5542">

**Comment by static-moonlight** [comment]
Thanks for the explanation.

It's still a little disappointing. I can't possibly be the only person, who thinks it might be a good idea to use the workflow message field to visualize why a workflow had failed, or can I? I don't understand why this is so difficult :(

**Reply by jswxstw** [reply]
You are not the only one, we actually have similar requirements, but from the implementation perspective of the controller, it seems quite difficult to achieve it. For example, the same workflow may have many failed nodes. How does the controller traverse during scheduling and determine which node's failure message to use?

Exit Handler can access the global variable `workflow.failures`, you can try to save it, but it may not be possible to write into the workflow message.

**Comment by janeisklar** [comment]
I was trying to achieve the same thing (on success) by passing a service account to the container which is then used to patch the workflow, e.g. something similar to: `kubectl patch workflow my-workflow-vjjm2 --type='json' --patch='[{"op":"replace","path":"/status/message","value":"test"}]'`.

I can see the message for a few seconds before the workflow goes from `Running` to `Succeeded` which will then eventually override it with `-`. Very disappointing:(

I understand that it may be hard to know which message to use, but if I'm setting one myself do you think it would be possible not to have it removed at the end if a value is already set?

**Reply by jswxstw** [reply]
> I can see the message for a few seconds before the workflow goes from Running to Succeeded which will then eventually override it with -. Very disappointing:(

You are right. That's why I said it's not possible to update the message in the exit handler. Either update it asynchronously when the workflow is completed?

> I understand that it may be hard to know which message to use, but if I'm setting one myself do you think it would be possible not to have it removed at the end if a value is already set?

https://github.com/argoproj/argo-workflows/blob/6201d759ebb32ba543c01665c142d04b0428f227/workflow/controller/operator.go#L2357-L2361

`workflow.message` should be the result of the latest execution after scheduling. If it is not overwritten and the old result is retained, this is somewhat counterintuitive.

**Reply by janeisklar** [reply]
@jswxstw thanks for the code reference!

I see, so on success the message is always overwritten with an empty string:(

https://github.com/argoproj/argo-workflows/blob/6201d759ebb32ba543c01665c142d04b0428f227/workflow/controller/operator.go#L2452-L2454

How would I go about setting it asynchronously? The workflow finishes only after my container has stopped, so I don't really see how I would be able to run something afterwards.

**Reply by jswxstw** [reply]
> How would I go about setting it asynchronously? The workflow finishes only after my container has stopped, so I don't really see how I would be able to run something afterwards.

For example, introduce a new component, implement a workflow informer that listens for workflow update events, and upon workflow failed, search for the corresponding message of failed nodes and overwrite the `workflow.message`.

**Reply by janeisklar** [reply]
Thanks for the pointer, I'll look into it!

**Reply by jswxstw** [reply]
Perhaps you can propose an enhancement to add this functionality to the controller.

**Reply by janeisklar** [reply]
I unfortunately won't be able to share what I've built, but nevertheless wanted to outline my solution.

When running my workflow step I am annotating the current workflow which I identify using the environment variable `ARGO_NODE_ID`. I  thereby inject a service account (in my case by setting `cronworkflow.spec.workflowSpec.serviceAccountName`) which allows me to use the Kubernetes API to annotate the workflow with the name `ARGO_NODE_ID`. The value of the annotation I set to the message I would like to be visible upon job completion.

Secondly, I specify a `MutatingWebhookConfiguration` which will executed when changes are made to the Workflow state:
```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: MutatingWebhookConfiguration
metadata:
  annotations:
    cert-manager.io/inject-ca-from: {{ .Release.Namespace }}/{{ $shortName }}
  name: {{ include "helm.name" . }}-webhook
webhooks:
- admissionReviewVersions:
  - v1
  clientConfig:
    service:
      name: {{ include "helm.fullname" . }}
      namespace: {{ .Release.Namespace }}
      path: /mutate
      port: {{ .Values.service.port }}
  failurePolicy: Fail
  matchPolicy: Equivalent
  name: myname
  reinvocationPolicy: Never
  rules:
    - apiGroups:
        - argoproj.io
      apiVersions:
        - v1alpha1
      operations:
        - UPDATE
      resources:
        - workflows
      scope: '*'
  sideEffects: None
  timeoutSeconds: 3
```

The webhook will accept all changes made to the workflow, except when the state changes to Succeeded/Failed and the annotation is present on the workflow. In that case the webhook responds with a patch that will set the workflow's `/status/message` to the value of the annotation:
```java
        var patchArray = objectMapper.createArrayNode();
        
        patchArray
            .addObject()
            .put("op", "replace")
            .put("path", "/status/message")
            .put("value", message);
```

It's maybe not the most straightforward way to solve this, but it is working nicely for me and I can see the message in the overview both on success and failure.

Having said all that, I absolutely agree that it would be better to actually solve this in Argo Workflows itself. However, coming from Java/Rust I don't have the Go skill to provide a pull request for this. Nevertheless I hope this workaround might help someone desperate to work around this issue.
